File: model-runner/python/metrics.py
"""Stream metrics back to the OpenModelStudio API in real-time."""

import logging
import threading
import time
from datetime import datetime, timezone
from typing import Optional, List, Dict, Any

import requests

logger = logging.getLogger("openmodelstudio.metrics")


class MetricReporter:
    """Buffers and streams metrics to the internal API endpoint.

    Posts individual MetricEvent objects to POST /internal/metrics/{job_id}.
    """

    def __init__(self, metrics_endpoint: str, job_id: str, flush_interval: float = 2.0):
        self.metrics_endpoint = metrics_endpoint.rstrip("/")
        self.job_id = job_id
        self.flush_interval = flush_interval
        self._buffer: List[Dict[str, Any]] = []
        self._lock = threading.Lock()
        self._running = True

        if self.metrics_endpoint:
            self._thread = threading.Thread(target=self._flush_loop, daemon=True)
            self._thread.start()
        else:
            self._thread = None
            logger.warning("No METRICS_ENDPOINT set, metrics will only be logged locally")

    def log(
        self,
        name: str,
        value: float,
        step: Optional[int] = None,
        epoch: Optional[int] = None,
    ):
        logger.info(f"metric: {name}={value} step={step} epoch={epoch}")
        entry = {
            "metric_name": name,
            "value": float(value),
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        if step is not None:
            entry["step"] = int(step)
        if epoch is not None:
            entry["epoch"] = int(epoch)
        with self._lock:
            self._buffer.append(entry)

    def _flush_loop(self):
        while self._running:
            time.sleep(self.flush_interval)
            self._flush()

    def _flush(self):
        with self._lock:
            if not self._buffer:
                return
            batch = self._buffer[:]
            self._buffer.clear()

        if not self.metrics_endpoint:
            return

        url = f"{self.metrics_endpoint}/{self.job_id}"
        failed = []
        for entry in batch:
            try:
                resp = requests.post(url, json=entry, timeout=10)
                if resp.status_code >= 400:
                    logger.warning(f"Metrics POST returned {resp.status_code}: {resp.text[:200]}")
                    failed.append(entry)
            except Exception as e:
                logger.warning(f"Failed to post metric: {e}")
                failed.append(entry)

        if failed:
            with self._lock:
                self._buffer = failed + self._buffer

    def close(self):
        self._running = False
        self._flush()
        if self._thread:
            self._thread.join(timeout=5)

File: model-runner/rust/src/metrics.rs
use serde::Serialize;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Serialize)]
pub struct MetricEntry {
    pub name: String,
    pub value: f64,
    pub timestamp: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub step: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub epoch: Option<i64>,
}

pub struct MetricReporter {
    api_url: String,
    job_id: String,
    buffer: Arc<Mutex<Vec<MetricEntry>>>,
}

impl MetricReporter {
    pub fn new(api_url: String, job_id: String) -> Self {
        Self {
            api_url,
            job_id,
            buffer: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn log(&self, name: &str, value: f64, step: Option<i64>, epoch: Option<i64>) {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let entry = MetricEntry {
            name: name.to_string(),
            value,
            timestamp,
            step,
            epoch,
        };

        let mut buf = self.buffer.lock().unwrap();
        buf.push(entry);

        // Auto-flush at 50 entries
        if buf.len() >= 50 {
            let batch: Vec<_> = buf.drain(..).collect();
            drop(buf);
            let _ = self.flush_batch(batch);
        }
    }

    fn flush_batch(&self, batch: Vec<MetricEntry>) -> Result<(), Box<dyn std::error::Error>> {
        let client = reqwest::blocking::Client::new();
        let url = format!("{}/api/jobs/{}/metrics", self.api_url, self.job_id);
        client.post(&url)
            .json(&serde_json::json!({"metrics": batch}))
            .timeout(std::time::Duration::from_secs(10))
            .send()?;
        Ok(())
    }

    pub fn flush(&self) {
        let mut buf = self.buffer.lock().unwrap();
        if buf.is_empty() { return; }
        let batch: Vec<_> = buf.drain(..).collect();
        drop(buf);
        let _ = self.flush_batch(batch);
    }
}

impl Drop for MetricReporter {
    fn drop(&mut self) {
        self.flush();
    }
}
